### PR TITLE
Make CheckType more functional

### PR DESCRIPTION
**Pull request by bryancroteau-MSFT**
Small change to CheckType in Binder.cs to move the main logic to a static helper, which is more functional (a mapping of inputs to outputs).

### CODE REVIEW

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/libraries/Microsoft.PowerFx.Core/Binding/BinderCheckTypeResult.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Syntax;
+
+namespace Microsoft.PowerFx.Core.Binding
+{
+    internal class BinderCheckTypeResult
+    {
+        public bool Success { get; init; }
+
+        public IReadOnlyCollection<BinderCoercionResult> Coercions { get; init; }
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Binding/BinderCheckTypeResult.cs:14]: comment about how this is used?

File: src/libraries/Microsoft.PowerFx.Core/Binding/BinderCheckTypeResult.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System.Collections.Generic;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Syntax;
+
+namespace Microsoft.PowerFx.Core.Binding
+{
+    internal class BinderCheckTypeResult
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Binding/BinderCheckTypeResult.cs:10]: Could these reasonably be `readonly struct`? We haven't really been using this yet in PowerFx, but this seems suited to it. 
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/struct#readonly-struct

**Review by lesaltzm — APPROVED** [review]
minor comments, approved

File: src/libraries/Microsoft.PowerFx.Core/Binding/BinderCheckTypeResult.cs (hunk 1 above)
Comment (bryancroteau-MSFT) [reply src/libraries/Microsoft.PowerFx.Core/Binding/BinderCheckTypeResult.cs:17]: Done

File: src/libraries/Microsoft.PowerFx.Core/Binding/BinderCheckTypeResult.cs (hunk 2 above)
Comment (bryancroteau-MSFT) [reply src/libraries/Microsoft.PowerFx.Core/Binding/BinderCheckTypeResult.cs]: Done

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Review by lesaltzm — APPROVED** [review]
🚢

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Pull request merged**